src/sign_outgoing.rs

```diff
@@ -0,0 +1,112 @@
+use crate::MAuthInfo;
+use chrono::prelude::*;
+use reqwest::{header::HeaderValue, Request};
+use thiserror::Error;
+
+impl MAuthInfo {
+    /// This method determines how to sign the request automatically while respecting the
+    /// `v2_only_sign_requests` flag in the config file. It always signs with the V2 algorithm and
+    /// signature, and will also sign with the V1 algorithm, if the configuration permits.
+    ///
+    /// Note that, as the request signature includes a timestamp, the request must be sent out
+    /// shortly after the signature takes place.
+    ///
+    /// Note that it will need to read the entire body in order to sign it, so it will not
+    /// work properly if any of the streaming body types are used.
+    pub fn sign_request(&self, req: &mut Request) -> Result<(), SigningError> {
+        self.sign_request_v2(req)?;
+        if self.sign_with_v1_also {
+            self.sign_request_v1(req)?;
+        }
+        Ok(())
+    }
+
+    /// Sign a provided request using the MAuth V2 protocol. The signature consists of 2 headers
+    /// containing both a timestamp and a signature string, and will be added to the headers of the
+    /// request. It is required to pass a `body_digest` computed by the
+    /// [`build_body_with_digest`](#method.build_body_with_digest) method, even if the request is
+    /// an empty-body GET.
+    ///
+    /// Note that, as the request signature includes a timestamp, the request must be sent out
+    /// shortly after the signature takes place.
+    ///
+    /// Also note that it will need to read the entire body in order to sign it, so it will not
+    /// work properly if any of the streaming body types are used.
+    pub fn sign_request_v2(&self, req: &mut Request) -> Result<(), SigningError> {
+        let timestamp_str = Utc::now().timestamp().to_string();
+        let body_data = match req.body() {
+            None => &[],
+            Some(reqwest_body) => reqwest_body.as_bytes().unwrap_or(&[]),
+        };
+        let some_string = self.signer.sign_string(
+            2,
+            req.method().as_str(),
+            req.url().path(),
+            req.url().query().unwrap_or(""),
+            body_data,
+            timestamp_str.clone(),
+        )?;
+        self.set_headers_v2(req, some_string, &timestamp_str);
+        Ok(())
+    }
+
+    pub(crate) fn set_headers_v2(&self, req: &mut Request, signature: String, timestamp_str: &str) {
```

Comment:
does this need to be available in the whole crate?

Reply:
Basically yes, the protocol test suite tests need access to it to test that the headers get set correctly. I'm not sure if there's a good way to provide access just to that.